File: src/de/kimminich/kata/tcg/Player.java
package de.kimminich.kata.tcg;

import de.kimminich.kata.tcg.strategy.Strategy;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.logging.Logger;

public class Player {

    private static final Logger logger = Logger.getLogger(Game.class.getName());

    private static final int STARTING_HAND_SIZE = 3;

    protected int health = 30;

    protected int manaSlots = 0;
    protected int mana = 0;

    protected List<Card> deck = Card.list(0, 0, 1, 1, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 5, 5, 6, 6, 7, 8);
    protected List<Card> hand = new ArrayList<>();

    protected final RandomCardPicker cardPicker;
    private final Strategy strategy;
    private final String name;

    public Player(String name, RandomCardPicker cardPicker, Strategy strategy) {
        this.name = name;
        this.cardPicker = cardPicker;
        this.strategy = strategy;
    }

    public int getHealth() {
        return health;
    }

    public int getMana() {
        return mana;
    }

    public int getNumberOfDeckCardsWithManaCost(int manaCost) {
        return (int) deck.stream().filter(card -> card.getManaCost() == manaCost).count();
    }

    public int getNumberOfDeckCards() {
        return deck.size();
    }

    public Integer getNumberOfHandCardsWithManaCost(int manaCost) {
        return (int) hand.stream().filter(card -> card.getManaCost() == manaCost).count();
    }

    public int getNumberOfHandCards() {
        return hand.size();
    }

    public void drawCard() {
        if (getNumberOfDeckCards() == 0) {
            logger.info(this + " bleeds out!");
            health--;
        } else {
            Card card = cardPicker.pick(deck);
            deck.remove(card);
            logger.info(this + " draws card: " + card);
            if (getNumberOfHandCards() < 5) {
                hand.add(card);
            } else {
                logger.info(this + " drops card " + card + " from overload!");
            }
        }
    }

    public int getManaSlots() {
        return manaSlots;
    }

    public void giveManaSlot() {
        manaSlots++;
    }

    public void refillMana() {
        mana = manaSlots;
    }

    public void drawStartingHand() {
        for (int i = 0; i < STARTING_HAND_SIZE; i++) {
            drawCard();
        }
    }

    public void playCard(Card card, Player opponent) {
        if (mana < card.getManaCost()) {
            throw new IllegalMoveException("Insufficient Mana (" + mana + ") to pay for card " + card + ".");
        }
        logger.info(this + " plays card: " + card);
        mana -= card.getManaCost();
        hand.remove(card);
        opponent.receiveDamage(card.getDamage());
    }

    private void receiveDamage(int damage) {
        health -= damage;
    }

    public boolean canPlayCards() {
        return hand.stream().filter(card -> card.getManaCost() <= mana).count() > 0;
    }

    public void playCard(Player opponent) {
        Optional<Card> card = strategy.nextCard(mana, hand);
        if (card.isPresent()) {
            playCard(card.get(), opponent);
        } else {
            throw new IllegalMoveException("No card can be played from hand " + hand + " with (" + mana + ") mana.");
        }
    }

    @Override
    public String toString() {
        return "Player:" + name + "{" +
                "health=" + health +
                ", mana=" + mana + "/" + manaSlots +
                ", hand=" + hand +
                ", deck=" + deck +
                '}';
    }

}

File: tcg-kotlin/src/main/kotlin/de/kimminich/kata/tcg/Player.kt
package de.kimminich.kata.tcg

import de.kimminich.kata.tcg.custom.IllegalMoveException
import de.kimminich.kata.tcg.strategy.Strategy
import java.util.*
import kotlin.math.min

class Player {
    private var random = Random()
    var health = MAXIMUM_HEALTH
        private set
    private var manaSlots = 0
    private var mana = 0
    private var deck: MutableList<Card> = createListOfCards(0, 0, 1, 1, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 5, 5, 6, 6, 7, 8)
    private var hand: MutableList<Card> = ArrayList()
    private val strategy: Strategy
    private val name: String

    constructor(name: String, strategy: Strategy) {
        this.name = name
        this.strategy = strategy
    }



    internal constructor(
        name: String,
        strategy: Strategy,
        health: Int,
        manaSlots: Int,
        mana: Int,
        deck: MutableList<Card>,
        hand: MutableList<Card>
    ) {
        this.name = name
        this.strategy = strategy
        this.health = health
        this.manaSlots = manaSlots
        this.mana = mana
        this.deck = deck
        this.hand = hand
    }

    fun getNumberOfDeckCardsWithManaCost(manaCost: Int): Int {
        return deck.stream().filter { card: Card -> card.value == manaCost }.count().toInt()
    }

    val numberOfDeckCards: Int
        get() = deck.size

    fun getNumberOfHandCardsWithManaCost(manaCost: Int): Int {
        return hand.stream().filter { card: Card -> card.value == manaCost }.count().toInt()
    }

    val numberOfHandCards: Int
        get() = hand.size

    fun drawCard() {
        if (numberOfDeckCards == 0) {
            println("$this bleeds out!")
            health--
        } else {
            val card = deck[random.nextInt(deck.size)]
            deck.remove(card)
            println("$this draws card: $card")
            if (numberOfHandCards < MAXIMUM_HAND_SIZE) {
                hand.add(card)
            } else {
                println("$this drops card $card from overload!")
            }
        }
    }

    fun giveManaSlot() {
        if (manaSlots < MAXIMUM_MANA_SLOTS) {
            manaSlots++
        }
    }

    fun refillMana() {
        mana = manaSlots
    }

    fun drawStartingHand() {
        for (i in 0..<STARTING_HAND_SIZE) {
            drawCard()
        }
    }

    private fun heal(amount: Int) {
        health = min((health + amount).toDouble(), MAXIMUM_HEALTH.toDouble()).toInt()
    }

    private fun receiveDamage(damage: Int) {
        health -= damage
    }

    fun canPlayCards(): Boolean {
        return hand.stream().filter { card: Card -> card.value <= mana }.count() > 0
    }

    fun playCard(opponent: Player) {
        val move = strategy.nextMove(mana, health, hand)
        playCard(move.card, opponent, move.action)
    }

    fun playCard(card: Card, opponent: Player, action: Action) {
        if (mana < card.value) {
            throw IllegalMoveException("Insufficient Mana ($mana) to pay for card $card.")
        }
        println("$this plays card $card for $action")
        mana -= card.value
        hand.remove(card)
        when (action) {
            Action.DAMAGE -> opponent.receiveDamage(card.value)
            Action.HEALING -> heal(card.value)
            else -> throw IllegalMoveException("Unrecognized game action: $action")
        }
    }

    override fun toString(): String {
        return "Player:" + name + "{" +
                "health=" + health +
                ", mana=" + mana + "/" + manaSlots +
                ", hand=" + hand +
                //", deck=" + deck +
                '}'
    }

    fun getManaSlots(): Int {
        return manaSlots
    }
    fun getMana(): Int {
        return mana
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Player

        return name == other.name
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }


    companion object {
        private const val STARTING_HAND_SIZE = 3
        private const val MAXIMUM_HAND_SIZE = 5
        private const val MAXIMUM_HEALTH = 30
        private const val MAXIMUM_MANA_SLOTS = 10
    }


}
